rig-core/src/agent/completion.rs

```diff
@@ -0,0 +1,251 @@
+                    .filter_map(|toolname| async move {
+                        if let Some(tool) = self.tools.get(toolname) {
+                            Some(tool.definition(text.into()).await)
+                        } else {
+                            tracing::warn!(
+                                "Tool implementation not found in toolset: {}",
+                                toolname
+                            );
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .await;
+
+                completion_request
+                    .documents(dynamic_context)
+                    .tools([static_tools.clone(), dynamic_tools].concat())
+            }
+            None => {
+                let static_tools = stream::iter(self.static_tools.iter())
+                    .filter_map(|toolname| async move {
+                        if let Some(tool) = self.tools.get(toolname) {
+                            // TODO: tool definitions should likely take an `Option<String>`
+                            Some(tool.definition("".into()).await)
+                        } else {
+                            tracing::warn!(
+                                "Tool implementation not found in toolset: {}",
+                                toolname
+                            );
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .await;
+
+                completion_request.tools(static_tools)
+            }
+        };
+
+        Ok(agent)
+    }
+}
+
+// Here, we need to ensure that usage of `.prompt` on agent uses these redefinitions on the opaque
+//  `Prompt` trait so that when `.prompt` is used at the call-site, it'll use the more specific
+//  `PromptRequest` implementation for `Agent`, making the builder's usage fluent.
+//
+// References:
+//  - https://github.com/rust-lang/rust/issues/121718 (refining_impl_trait)
+
+#[allow(refining_impl_trait)]
+impl<M: CompletionModel> Prompt for Agent<M> {
+    fn prompt(
+        &self,
+        prompt: impl Into<Message> + Send,
+    ) -> PromptRequest<M, prompt_request::Simple> {
+        PromptRequest::new(self, prompt)
+    }
+}
+
+#[allow(refining_impl_trait)]
+impl<M: CompletionModel> Prompt for &Agent<M> {
+    fn prompt(
+        &self,
+        prompt: impl Into<Message> + Send,
+    ) -> PromptRequest<M, prompt_request::Simple> {
+        PromptRequest::new(*self, prompt)
+    }
+}
+
+#[allow(refining_impl_trait)]
+impl<M: CompletionModel> Chat for Agent<M> {
+    async fn chat(
+        &self,
+        prompt: impl Into<Message> + Send,
+        chat_history: Vec<Message>,
+    ) -> Result<String, PromptError> {
+        let mut cloned_history = chat_history.clone();
+        PromptRequest::new(self, prompt)
+            .with_history(&mut cloned_history)
+            .await
+    }
+}
```

Comment:
The fact that the `PromptRequest` now accepts a chat history opens the question of whether or not the `Chat` trait is useful at all anymore :thinking:

Reply:
yeaaa, i actually want to deprecate it but i wasn't sure bc we use it as an interface in several examples. I think it'll be more natural for fancier multi-agent systems to use the tower, middle-ware interface rather than the `Chat` trait.

Reply:
The plan is to deprecate this in a future PR